### PR TITLE
DIG-824: Refactoring FL-Repo for easier experiment injection

### PR DESCRIPTION
**Pull request by AliRZ-02**
## Description
This PR deals with the [DIG-817](https://candig.atlassian.net/browse/DIG-817) and [DIG-824](https://candig.atlassian.net/browse/DIG-824) tickets. The former deals with "[Performing] differentially private federated learning experiment[s]", whereas the latter deals specifically with "Refactor[ing] [the] federated-learning repository to permit injection of custom differential privacy algorithms". This PR mostly changes the FL services to permit greater flexibility, such as additional dependencies and custom Strategies. It removes all experiment-specific code from fl-server and fl-client. We also add documentation on how to add your custom differential privacy algorithms, alongside some technical debt related to custom differential privacy algorithms. Finally, we ensure all experiment-specific code is stored in its own subdirectory within the larger `experiments` folder.

## Changelog
- Removed experiment-specific code from fl-server/fl-client
    - Removed extra dependencies from requirements.txt
        - to be injected instead
    - Removed Strategy code in fl-server
        - to be injected instead
    - Removed repeated code for Synthea DP experiment
        - uses quickstart to inject experiment files into a docker container, instead of repeating code
- Added Documentation on implementing Differentially-Private FL experiment/algorithm
- Improved privacy level of DP experiment
- Improved accuracy of non-DP Federated Synthea Experiment
- Adds MIT license

## Dependencies
- N/A

## Testing
- We've tested the changes on separate machines and encountered no problems

### CODE REVIEW

File: experiments/synthea-breast-cancer/winter2022/Differentially-Private/quickstart.sh [hunk 1]
```diff
@@ -30,13 +30,14 @@ help ()
...
-   echo "   -n <SITES>           Number of Sites to federate. Defaults to 2"
-   echo "   -r <ROUNDS>          Number of rounds of trials to conduct. Defaults to 100."
-   echo "   -e <EXPERIMENT_PATH> Pass in the path to the experiment folder.Defaults to ./experiment in the root folder. Ensure this path is either an absolute path, or that it starts with './'"
-   echo "   -s                   Keep all of the data in one dataset - Useful only if -i specified"
-   echo "   -h                   Display this help text"
+   echo "   -i <INGEST_PATH>               Ingest Data present at Path into Katsu"
+   echo "   -p <PORT>                      Specify Port Number to expose. Defaults to 5000"
+   echo "   -n <SITES>                     Number of Sites to federate. Defaults to 2"
+   echo "   -r <ROUNDS>                    Number of rounds of trials to conduct. Defaults to 100."
+   echo "   -e <EXPERIMENT_PATH>           Pass in the path to the experiment folder. Defaults to ./experiment in the root folder. Ensure this path is either an absolute path, or that it starts with './'"
```
Comment (yuriy-pavlov) [inline experiments/synthea-breast-cancer/winter2022/Differentially-Private/quickstart.sh:37]: The difference between the `EXPERIMENT_PATH` and `FEDERATED_EXPERIMENT_PATH` is not very clear; I don't think a new dev would understand what's happening here. Could you clarify (and maybe rename these variable names?) here and in the relevant documentation in `./README.md`?

File: experiments/synthea-breast-cancer/winter2022/Differentially-Private/quickstart.sh [hunk 2]
```diff
@@ -250,6 +259,35 @@ if [[ ${TO_INGEST} -eq 1 ]]; then
     sleep ${SLEEP_TIME}
 fi
 
+# Add Experiment to Containers
+echo
+docker-compose up -d
+
+## Kill All Running FL-* containers
```
Comment (yuriy-pavlov) [inline experiments/synthea-breast-cancer/winter2022/Differentially-Private/quickstart.sh:266]: The user should be informed that running `fl-*` containers are going to be killed. Add a note about this to the `help` for this script, and also have the script `echo` the fact that it is about to perform the kill.

File: orchestration-scripts/configure_docker_compose.py [hunk 3]
```diff
@@ -141,6 +142,14 @@ def save_file_here(contents: str, name: str) -> None:
     file.write(contents)
     print(f"Docker-Compose File saved at {os.getcwd()}/{name}!")
 
+def copy_experiment_requirements(experiment_path: str):
```
Comment (yuriy-pavlov) [inline orchestration-scripts/configure_docker_compose.py:145]: `configure_docker_compose.py` shouldn't be responsible for doing this, an action that modifies another part of the filesystem. While it's useful, it's an unexpected side-effect. Is there anywhere else we could easily move this action to ("easy" = 30 mins of work or less)? If not, add a note about this to the `Technical Debt` section at the bottom of the main README.

File: orchestration-scripts/remove_bind_mounts.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    Function to save a file called `name` in the `pwd` with the given `contents`.
+
+    Can't import from configure_docker_compose.py, because module with dash `orchestration-scripts` cannot be imported from with ease
+    """
+    
+    file = open(name, "w")
+    file.write(contents)
+    print(f"Modified Docker-Compose File saved at {os.getcwd()}/{name}!")
+
+def main() -> None:
```
Comment (yuriy-pavlov) [inline orchestration-scripts/remove_bind_mounts.py:70]: Thank you for adding this utility script.

File: services/fl-server/server.py [hunk 5]
```diff
@@ -1,56 +1,22 @@
...
-# Start Flower server for five rounds of federated learning
+# Start Flower server
 if __name__ == "__main__":
     experiment.set_initial_params(model)
-    strategy = SaveModelStrategy(
+    strategy = Strategy(
         min_available_clients=settings.FL_MIN_CLIENTS,
         eval_fn=get_eval_fn(),
-        on_fit_config_fn=fit_round,
+        on_fit_config_fn=lambda rnd: {"rnd": rnd} # Don't need named function to simply return round number in a dict instead of int
```
Comment (yuriy-pavlov) [inline services/fl-server/server.py:19]: Nice to see these small improvements to the code. Thank you for putting in the effort.
You can remove the inline comment (`Don't need named function...`), it was helpful for me to see as the reviewer but it won't be needed after review.

**Review by sh-ura — CHANGES_REQUESTED** [review]
LGTM; I requested a few small changes but it's ready to merge otherwise.

Good documentation of the differential-privacy implementation and its potential future improvements.

The refactor itself (abstraction of `Strategy`) is suitable and sufficient; good work.

Thanks for deleting the duplicated experiment code; grabbing it from the previous non-DP experiment is a good call, although the documentation for doing so is a little confusing.

File: orchestration-scripts/configure_docker_compose.py (hunk 3 above)
Comment (AliRZ-02) [reply orchestration-scripts/configure_docker_compose.py:145]: I've made a note of this within the technical debt section, but the reason we have the code within this file is that Python is really finicky with imports and can't import from any folder with a dash in its name. So we can't import from `orchestration-scripts` and thus we can't move the code into a new file without renaming the entire folder, which means we would have to change the filepath location everywhere else it is used as well.

**Review by sh-ura — APPROVED** [review]
Looks good, thank you!

**Pull request merged**